osquery/tables/system/ssh_keys.cpp

```diff
@@ -15,12 +15,115 @@
 #include <osquery/filesystem/filesystem.h>
 #include <osquery/logger/logger.h>
 #include <osquery/tables/system/system_utils.h>
+#include <osquery/utils/scope_guard.h>
 #include <osquery/utils/system/system.h>
 
+#include <boost/algorithm/string.hpp>
+#include <openssl/bio.h>
+#include <openssl/evp.h>
+#include <openssl/pem.h>
+
 namespace osquery {
 namespace tables {
 
-const std::string kSSHUserKeysDir = ".ssh/";
+extern const std::string kSSHUserKeysDir = ".ssh";
+
+const std::string kOpenSshHeader = "-----BEGIN OPENSSH PRIVATE KEY-----";
+
+// The first bytes of an OpenSSH key are |key type|length of cipher name|cipher
+// name| so all unencrypted ed25519 keys should start with the value below
+// (encoded in base64) This magic string is (hex-ified):
+//  6f 70 65 6e 73 73 68 2d 6b 65 79 2d 76 31 00 00 00 00 04 6e 6f 6e 65
+// |--------------------------------------------|-----------|----------|
+// | o  p  e  n  s  s  h  -  k  e  y  -  v  1 \0|          4| n  o  n e|
+const std::string kOpenSshUnencryptedPrefix = "b3BlbnNzaC1rZXktdjEAAAAABG5vbmU";
+
+// openssl can't currently parse the new openssh keys
+bool isOpenSSHKey(std::string& keys_content) {
+  return boost::starts_with(keys_content, kOpenSshHeader);
+}
+
+// `true` if the openssh key is encrypted, `false` otherwise.
+bool isOpenSSHKeyEncrypted(std::string& keys_content) {
+  const std::string prefix = keys_content.substr(
+      kOpenSshHeader.size() + 1, kOpenSshUnencryptedPrefix.size());
+  return prefix != kOpenSshUnencryptedPrefix;
+}
+
+// parsePrivateKey returns true iff the key is valid.
+// Tries to parse the .PEM using openssl. If that fails, it checks
+// if it's an openssh key.
+bool parsePrivateKey(std::string& keys_content,
+                     int* key_type,
+                     bool* is_encrypted) {
+  BIO* bio_stream = BIO_new(BIO_s_mem());
+  auto const bio_stream_guard =
+      scope_guard::create([bio_stream]() { BIO_free(bio_stream); });
+  BIO_write(bio_stream, keys_content.c_str(), keys_content.size());
+  if (bio_stream == nullptr) {
+    return false;
+  }
+
+  // PEM_read_bio_PrivateKey calls passwordCallback
+  // if the private key is encrypted. We don't care what the key is;
+  // only whether or not it's encrypted.
+  static bool encrypted;
```

Comment:
Why is this a `static bool` and not `bool`?

Reply:
So this is interesting. This was done by Robbie Ostrow who's code I revived to open this PR, so I had to do a little playing around to figure this out. 

The reason it's `static bool` instead of simply `bool`, is that Robbie was trying (as much as possible) to use OpenSSL to determine if the private key is encrypted, hence where he used `PEM_read_bio_PrivateKey`, which expects a callback function with a particular signature as it's third argument. 

The lambda that is constructed on line 72 can be converted to that callback, but only if it doesn't do any captures, e.g. 

```
  bool encrypted;
  encrypted = false;
  auto passwordCallback = [encrypted](char*, int, int, void*) {
    encrypted = true;
    return -1; // let openssl know that the passwordCallback failed
  };
```
  
  Which I did try. ^ That fails to compile because of the signature mismatch when passing in to `PEM_read_bio_PrivateKey`. 
  
  So Robbies solution was to make it a capture-less lambda, and make `encrypted` static so that it could be accessed from the lambda's scope. Alternately he could have made it a true function rather than a lambda but that would have had the same issue. 
   
So the main thing is: 
 a) Trying to use OpenSSL rather than hand-rolling functionality (which is good), but unfortunately they don't provide an "IsKeyEncrypted" function
 b) The function being used accepts a really specific callback rather than e.g. std::function that would be easier to customize
 
I'm happy to try to find a different solution if you want, just let me know.

Reply:
The last argument in the `PEM_read_bio_PrivateKey` function as with the callback that's provided it's a user specified value that can be passed to the callback itself.
This is often a way present to pass a user context; you can pass **the address** of the encrypted variable via `reinterpret_cast<void*>(encrypted)`, and then `reintepret_cast<bool*>(encrypted)` it back inside the callback.

Reply:
Et voilà! Thanks @Smjert , I've made that change now and `encrypted` is no longer static.